Show status information in terminal title

michaelkuhn opened the pull request:
Installing packages with a lot of dependencies does not have an easy way of judging the current progress (apart from running `spack spec -I pkg` in another terminal). This change makes Spack update the terminal's title with status information, including the current installation phase as well as information about the current and total number of packages.

Currently only tested on Fedora with GNOME Terminal (where it works :smile:).

adamjstewart commented:
Tested this on the builtin macOS Terminal app and it works great! Now if only I could get #11516 working reliably...

mwkrentel commented:
Could we make this a command-line option?  For me, I set my own title,
unset PROMPT_COMMAND and I really dislike it when programs reset it
against my wishes.

"Works great" is a personal preference here.

michaelkuhn commented:
> Could we make this a command-line option? For me, I set my own title,
> unset PROMPT_COMMAND and I really dislike it when programs reset it
> against my wishes.

Ah, so your title is not restored after Spack finishes because `PROMPT_COMMAND` is unset? I'll try to come up with a configuration option for it.

michaelkuhn commented:
I have pushed a new commit adding a configuration option `set_terminal_title` to `config.yaml`. Let me know whether this works for you.

mwkrentel commented:
I don't think PROMPT_COMMAND has anything to do with it.  I just meant
that this is a personal preference, not everyone likes it, so I normally
turn things like this off.

But actually, in your commit b56ff07f0942, I'm finding that it does
not reset the title.  Not with normal termination and not when I kill
it with Ctrl-C.  With normal termination, it ends with something like:

```
Spack: Running post-install hooks for bzip2 [3/3]
```

michaelkuhn commented:
> But actually, in your commit [b56ff07](https://github.com/spack/spack/commit/b56ff07f09421d949aaf77df3e392e25b45d4096), I'm finding that it does
> not reset the title. Not with normal termination and not when I kill
> it with Ctrl-C. With normal termination, it ends with something like:

Currently, I do not reset the title and let the shell take care of this (this is normally done when the shell shows the prompt after Spack has ended, for instance, via `PROMPT_COMMAND`). I don't think we can get the title before execution and restore it afterwards since there is no function to get the current title. We could only set it to something generic or clear it after the last package has finished installing.

mwkrentel commented:
I see, so I guess it is unsetting PROMPT_COMMAND that causes me trouble.
(I unset it because I didn't like cd changing the title, how ironic.)

As for enable/disable, I was thinking more of a command-line option
`spack -T/--title install package ...` but I can make config.yaml work.
Thanks!

I guess I'm the only one who doesn't like decorating the title.
There's old school and dinosaurs. :-)

Anyway, I'm good, so proceed.

michaelkuhn commented:
@tgamblin @alalazo Any opinions/feedback?

FYI: flatpak-builder is doing something similar.

michaelkuhn commented:
@alalazo Could you please review this? Thanks!

alalazo commented:
A couple of high-level questions:
1. How hard is it to use the last line on the screen (similarly to `apt`) instead of the terminal title?
2. How can we cope with parallel installations?

michaelkuhn commented:
> * How hard is it to use the last line on the screen (similarly to `apt`) instead of the terminal title?

Not sure, I have never tried to do something like this. I would imagine significantly harder, though, since we would have to redraw the last line every time something gets printed and make sure not to overwrite something.

> * How can we cope with parallel installations?

This version already works for parallel installations. If you do `spack install pkg` in two terminals, you get something like `Running install for pkg [1/15]` in the first terminal, `Running install for pkg [2/15]` in the second one etc. So we always get the total number of packages but simply skip the ones already being taken care of by a parallel install.

michaelkuhn commented:
@alalazo ping

alalazo commented:
@michaelkuhn I'd like other people to chime in on this. I have nothing against this PR, but being a matter of tastes it might be controversial - so I would like a broader consensus before merging. @tgamblin @becker33

becker33 commented:
I think at the very least we need to document this somewhere if we're going to merge it.

mwkrentel commented:
Note that this patch doesn't reset the title at the end of spack
install.  It relies on the user using PROMPT_COMMAND to do this.

I'm sure I'm in the minority here, but I set the title once and don't
use PROMPT_COMMAND.  (I'm commonly logged into 10-15 remote machines,
so I use the title for the name of the machine.)  Plus, most of the
time, I tee the output to a file (or it's from a cron job), and I
don't need all the extra control characters in the log.

That's why I suggested a command-line option like `spack install -T/--title`.
The option is really only useful if you're running interactively.

Or, maybe we could change `set_terminal_title` to 3-values: always,
never, or command-line.  I think everyone could find an acceptable
value in that case, including me.  :-)

But as long as there's some way to turn it off, don't let me get in the way.

michaelkuhn commented:
> Note that this patch doesn't reset the title at the end of spack
> install. It relies on the user using PROMPT_COMMAND to do this.

It's not really possible to reset the terminal title as far as I know. We could of course set it to something generic like the current directory but not to whatever the terminal has set before.

> I'm sure I'm in the minority here, but I set the title once and don't
> use PROMPT_COMMAND. (I'm commonly logged into 10-15 remote machines,
> so I use the title for the name of the machine.) Plus, most of the
> time, I tee the output to a file (or it's from a cron job), and I
> don't need all the extra control characters in the log.

The control characters are stripped already and should not end up in any log. Do you end up with any?

> That's why I suggested a command-line option like `spack install -T/--title`.
> The option is really only useful if you're running interactively.
> Or, maybe we could change `set_terminal_title` to 3-values: always,
> never, or command-line. I think everyone could find an acceptable
> value in that case, including me. :-)

Setting the terminal title is actually already skipped if run non-interactively (by checking `isatty`). I just fixed a small bug where some status updates were still set but this should not happen anymore.

I also added some documentation for the configuration option.

mwkrentel commented:
Ok, fair enough, you're right.  My normal use case is as follows, and
in this case, I don't get any stray control chars (it's a pipe, not a tty).

```
spack install ... 2>&1 | tee out.spack
```

@michaelkuhn  You really don't like the idea of a command-line option?
That would be my first thought.

Anyway, although it's not a feature that I care to use, I have
adequate ways to turn it off.  So, I don't have any objection.

michaelkuhn commented:
> @michaelkuhn You really don't like the idea of a command-line option?
> That would be my first thought.

I'm not really a fan of arguments you either have specify all the time or never to get the behavior you want. I'm already slightly annoyed by the amount of arguments I have to specify for other commands to make them behave like I want (`install -v`, `spec -I` etc.). Of course, I'd still rather be able to enable it with an argument than not at all.

Edit: We could also just disable it by default, then people could turn it on via, e.g., their user-scope `config.yaml`.

mwkrentel commented:
This is outside the scope of this PR, but ....  maybe what you really
want is a new config file to specify flags to various spack commands,
things like `install -v`, or `spec -I`, or `find -v -l`, etc.
(IMO, `install -v` would be a bad default, but if it's what you want.)

Or, maybe someone has already suggested this, I don't know.

alalazo commented:
@michaelkuhn Is this something you still pursue?

michaelkuhn commented:
Sorry for dropping the ball on this. Yes, this is still something I would like to see Spack support. I am planning to disable this feature by default and add a new parameter for `install` like @mwkrentel suggested (and then use the hopefully-soon-to-be-merged alias support to make it more convenient to use).

michaelkuhn committed:
installer: Support showing status information in terminal title

Installing packages with a lot of dependencies does not have an easy way
of judging the current progress (apart from running `spack spec -I pkg`
in another terminal). This change allows Spack to update the terminal's
title with status information, including its current progress as well as
information about the current and total number of packages.

michaelkuhn commented:
OK, I managed to update this faster than expected. It should now work with current `develop`. I made the following changes:
- Setting the terminal title is now disabled by default
- It can be activated on demand using `install -t`

When activated, it looks like this:
![term_title](https://user-images.githubusercontent.com/4062614/121745129-49005b80-cb04-11eb-8c8e-1761ff5356e0.png)

becker33 commented on lib/spack/spack/cmd/install.py:
```diff
@@ -378,6 +381,9 @@ def get_tests(specs):
     if args.deprecated:
         spack.config.set('config:deprecated', True, scope='command_line')
 
+    if args.terminal_title:
+        spack.config.set('config:set_terminal_title', True, scope='command_line')
```
There should be an equivalent option to turn off titles if the config default is turned on

michaelkuhn replied:
I added a `--no-terminal-title` option to turn it off explicitly. Not sure if this is what you meant.

becker33 commented on lib/spack/spack/installer.py:
```diff
@@ -683,6 +690,18 @@ def __str__(self):
+            return
+
+        status = '{0} [{1}/{2}]'.format(text, self.pkg_num, self.pkg_count)
+        print('\033]0;Spack: {0}\007'.format(status), end='')
```
Why use this instead of `sys.stdout.write`? Especially considering the number of sys.stdout calls necessary in this method.

michaelkuhn replied:
Good question. :smile: I probably started out with `print` and didn't think about changing it when the other `sys.stdout` calls became necessary. Fixed now.

becker33 commented:
@mwkrentel I have an idea for a follow-on feature that could be useful for your use case -- would you use a feature like this if there were a configurable option for how Spack should reset the title when it's done?

mwkrentel commented:
> would you use a feature like this if there were a configurable option
> for how Spack should reset the title when it's done?

@becker33 For me, not really.  Actually, I thought that PROMPT_COMMAND
already does what you're suggesting.  The ironic thing is that
unsetting PROMPT_COMMAND is what causes me trouble here.

Anyway, I don't want to get in the way here.  As long as there's a
simple way to turn this off, then I can live with it.  I just don't
want to have to say, `spack install --no-title ...` on every command.

On the plus side, I did figure out the control sequence to send to an
xterm to reset the title.  :-)

michaelkuhn commented:
I have resolved a conflict and rebased this to current develop. This should be good to go now.

sethrj commented on lib/spack/spack/installer.py:
```diff
@@ -693,6 +699,17 @@ def __str__(self):
         return '{0}: {1}; {2}; {3}; {4}'.format(
             self.pid, requests, tasks, installed, failed)
 
+    def _set_term_title(self, text, force=False):
```
The `force` isn't used by any caller. Was it meant for unit tests? (I don't see any actually.)

michaelkuhn replied:
That's actually just something I forgot to remove when updating the PR in the past. It's not necessary anymore.

sethrj commented on lib/spack/spack/installer.py:
```diff
@@ -693,6 +699,17 @@ def __str__(self):
+
+        status = '{0} [{1}/{2}]'.format(text, self.pkg_num, self.pkg_count)
+        sys.stdout.write('\033]0;Spack: {0}\007'.format(status))
+        sys.stdout.flush()
```
I think this functionality would be better as a small helper class rather than as a method with class data.  Maybe something like:
```python
class TermWriter(object):
    def __init__(self, pkg_count):
        self.pkg_num = 0
        self.pkg_count = pkg_count

    def next_package(self):
        self.pkg_num += 1

    def set_title(self, text):
        status = '{0} [{1}/{2}]'.format(text, self.pkg_num, self.pkg_count)
        sys.stdout.write('\033]0;Spack: {0}\007'.format(status))
        sys.stdout.flush()


class NullopTermWriter(object):
    def next_package(self):
        pass

    def set_title(self):
        pass


def make_term_writer(pkg_count):
    if (spack.config.get('config:set_terminal_title', False)
        and sys.stdout.isatty()):
        return TermWriter(pkg_count)

    # Terminal output disabled
    return NullopTermWriter()
```
the instantiate locally within `install`. It'll improve the ability to test and reduces class clutter.

michaelkuhn replied:
I've extracted it into a helper class now but kept it as a single class (I think we can afford the `if` within `set`).

fangohr commented:
I like the feature, and would like to see it available in spack.

sethrj commented on lib/spack/spack/installer.py:
```diff
@@ -1476,7 +1497,11 @@ def install(self):
         failed_explicits = []
         exists_errors = []
 
+        self.term_title = TermTitle(len(self.build_pq))
```
Does this need to be member data rather than being local to the function?

michaelkuhn replied:
You are right, a local variable is enough. Fixed.

haampie commented on lib/spack/spack/cmd/install.py:
```diff
@@ -132,6 +132,12 @@ def setup_parser(subparser):
     subparser.add_argument(
         '-v', '--verbose', action='store_true',
         help="display verbose build output while installing")
+    subparser.add_argument(
+        '-t', '--terminal-title', action='store_true', dest='terminal_title',
+        default=None, help="update terminal title while installing")
+    subparser.add_argument(
+        '--no-terminal-title', action='store_false', dest='terminal_title',
+        help="do not update terminal title while installing")
```
Do we need this? I can imagine `-t` can serve other purposes at some point, and this is going to be effectively the same as `spack -c config:set_terminal_title:false ...` (when that's supported)

michaelkuhn replied:
Do you mean just the short argument or the arguments in general?

Edit: Just saw your later comment. People have explicitly requested this to be a command-line argument before (see above).

haampie replied:
Hm yeah, but I noticed this comment:

> Anyway, I don't want to get in the way here. As long as there's a
> simple way to turn this off, then I can live with it. I just don't
> want to have to say, spack install --no-title ... on every command.

The feature is off by default, so, there's no reason for @mwkrentel to provide any flags? Otherwise do `spack config edit config` and set it to off? `spack -c config:terminal_title:false install` is of course terribly verbose, but I doubt you ever use this on a single invocation of `spack install` anyways... If you don't like it you'd want to disable it in user config once and for all through `spack config edit config`.

haampie replied:
Or `spack config add config:terminal_title:false` (pending a PR somewhere :p)

michaelkuhn replied:
The feature used to be on by default and @mwkrentel wanted an easy way to turn it off. Now it's off by default and `-t` provides an easy way to turn in on (and in combination with #17229 you can define an alias like `inst` for `install -tv`).

haampie replied:
Ah okay. Do you think you would actually use that command line option when you also have the config option? To me a command line option only makes sense if there is a need to enable/disable a feature for a single invocation of `spack install`, and I don't see in what circumstances you would enable `terminal_title` just once. But I may be missing something? Running multiple `spack install`'s in the background?

What do others think? My issue is that `-t` is likely to be rarely used, so it takes away a short-hand for [shiny new feature]. And with `--terminal-title` and `--no-terminal-title` I'm just afraid it's going to overwhelm the user when they look at `spack help` or tab completion: :sweat_smile: 

```
-h --help --only -u --until -j --jobs --overwrite --fail-fast --keep-prefix --keep-stage --dont-restage --use-cache --no-cache --cache-only --monitor --monitor-save-local --monitor-no-auth --monitor-tags --monitor-keep-going --monitor-host --monitor-prefix --include-build-deps --no-check-signature --require-full-hash-match --show-log-on-error --source -n --no-checksum --deprecated -v --verbose -t --terminal-title --no-terminal-title --fake --only-concrete --no-add -f --file --clean --dirty --test --run-tests --log-format --log-file --help-cdash --cdash-upload-url --cdash-build --cdash-site --cdash-track --cdash-buildstamp -y --yes-to-all
```

michaelkuhn replied:
I could also live without the `install` arguments as I will turn it on all the time anyway. But I will also define an alias for `install -v` as soon as #17229 is merged and would probably include `-t` for good measure.

tl;dr: I don't care about the configuration specifics, I just want to *finally* have useful progress reporting.

haampie replied:
If you don't introduce the command line args, leave the default to off in config.yaml, then we can probably just merge, since @sethrj already approved and so would I, and @mwkrentel should be happy too as the feature is opt-in.

The bike-shedding about command line args shouldn't be a blocker for the feature, so we can move that to another PR?

michaelkuhn replied:
Sounds good, done!

haampie commented on etc/spack/defaults/config.yaml:
```diff
@@ -190,3 +190,8 @@ config:
+  # Whether to set the terminal title to display status information during
+  # building and installing packages. This gives information about Spack's
+  # current progress as well as the current and total number of packages.
+  set_terminal_title: false
```
```suggestion
  terminal_title: false
```

shorter and more in line with `locks: true`, `ccache: false` etc.

michaelkuhn replied:
Well, we also have options with verbs like `verify_ssl` and `allow_sgid`, but I don't have any preference.

haampie reviewed (changes requested):
I like the feature, but just two bike-shedding comments:
1. I don't think we should reserve command line flags for this feature, because the same can be achieved by `-c` (maybe not now, but soon :tm:)
2. `terminal_title` true/false with `set_`?
